File: Python/binary_search.py
def binary_search(nums, target):

    """
    Performs binary search on a sorted input array to find the index of a target value.
    If the target value is not found, returns -1.

    Input:
    nums: Sorted input array
    target: Target value searching for its index

    Output:
    mid: index of "target" in "nums". If "nums" does not contain "target", mid=-1
    """

    # Check if the array is sorted
    if not all(nums[i] <= nums[i+1] for i in range(len(nums)-1)):
        raise ValueError("Input array must be sorted in ascending order")
    
    # Binary search
    low, high = 0, len(nums)-1
    while low <= high:
        mid = (low + high) // 2
        if target == nums[mid]:
            return mid
        elif target > nums[mid]:
            low = mid + 1
        else:
            high = mid - 1
    
    # Target value not found in array
    mid = -1
    return mid

if __name__ == "__main__":

    nums = [-1,0,1,3,5,4]
    target = 5

    print(binary_search(nums,target))

File: C++/binary_search.cpp
#include <iostream>
using namespace std;
#include <vector>

int binary_search(vector<int> nums, int target){
    /*
    Searches for integer "target" in vector "nums"
    Input:
        nums: sorted vector of integers
        target: integer to be searched for
    Output:
        ans: index of the integer "target"
    */


    // Check if the array is sorted
    for (int i = 1; i < nums.size(); i++) {
        if (nums[i] < nums[i-1]) {
            throw invalid_argument("Input array must be sorted in ascending order");
        }
    }

    // perform binary search
    int low = 0;
    int high = nums.size() - 1;
    int mid = (low + high)/2;

    while (low < high){
        if(nums[mid] < target){
            low = mid +1;
        }
        else{
            high = mid;
        }
        mid = (low + high)/2;
    }
    
    int ans = low;
    
    // check if target is in nums, if not return -1
    if (nums[ans] != target) {
        return -1;
    }

    return ans;
};

int main(){

    vector<int> nums = {0,1,5,7,9,10};
    int target = 10;

    int ans = binary_search(nums,target); 
    cout << "Index is: " << ans << endl;

    return 0;
}
